hibernate-core/src/main/antlr/org/hibernate/grammars/hql/HqlParser.g4

```diff
@@ -677,6 +670,14 @@ countFunction
 	: COUNT LEFT_PAREN DISTINCT? (expression | ASTERISK) RIGHT_PAREN
 	;
 
+everyFunction
+	: (EVERY|ALL) LEFT_PAREN (predicate | subQuery) RIGHT_PAREN
+	;
+
+anyFunction
+	: (ANY|SOME) LEFT_PAREN (predicate | subQuery) RIGHT_PAREN
+	;
+
```

Comment:
Note that `any`/`some` and `every`/`all` have quite different semantics depending upon whether their argument is a predicate or a subquery. However, I failed to convince myself that we really needed to distinguish these two roles in the parser.

Reply:
I'm having a hard time envisioning a situation for the predicate-form.  For example, what does `... where someLhsExpr > all (1=1)`.  Yes, I had predicate there initially but that was just wrong from the first drafts of the grammar because these qualifiers had not yet been tested.

Not specific to this PR, but I do not like the categorization of these as functions.  They are really more a qualifier or restriction of the sub-query-based predicate.  In terms of modeling I think it makes the most sense to wrap the sub-query in a ANY/SOME expression wrapper

Reply:
It's used as an aggregate function:

    select every(person.age>10) from Person p
    select any(order.status = Status.UNPAID) from Customer p where p.id = :id

Reply:
> Not specific to this PR, but I do not like the categorization of these as functions. They are really more a qualifier or restriction of the sub-query-based predicate.

Yeah well that's what I was alluding to when I said I failed to convince myself that we need to separate them in the *grammar*. The aggregate functions `every()` and `any()` are certainly functions and belong alongside `sum()` and friends.

But the subquery *quantifiers* aren't functions at all. (Indeed, I find them to be a rather strange syntax.)

Reply:
I've never seen that form before.  Nice!

Ok, so you mean to handle the "aggregate function" form via SQM functions and the sub-query qualifier form via `SqmAny`, etc?

Reply:
Ok, I see that is what you do in `SemanticQueryBuilder#visitAnyFunction`.  That's fine.  Personally I'd be inclined to define distinct rules just to be clear about the distinction, but I'm certainly not going to fight over it ;)

Reply:
> I've never seen that form before. Nice!

So it's actually part of ANSI SQL, if I'm reading the spec correctly, but I only found one database that supported them. However, it's possible to emulate them on most databases, so that's what I did (because I think they're useful).

Reply:
> Ok, so you mean to handle the "aggregate function" form via SQM functions and the sub-query qualifier form via `SqmAny`, etc?

Precisely.